### ISSUE TITLE
Regular segfaults

### ISSUE THREAD

**Issue by ejholmes** [draios/falco#299]
I've been noticing that some falco processes in our cluster have been regularly crashing. Here's a log line from `/var/log/syslog` with the crash:

```
<6>1 2017-11-16T07:24:44.059104+00:00 ip-10-128-18-204 kernel - - [facility=kern sev=info] [62400.157197] falco[4750]: segfault at 48 ip 000000000057bd13 sp 00007ffe3bd0c560 error 4 in falco[400000+623000]
```

I haven't yet pinpointed exactly what seems to be the source of it, but it's extra annoying since the [installed init script](https://github.com/draios/falco/blob/dev/scripts/debian/falco) doesn't seem to successfully restart it, and falco remains stopped until someone manually restarts it.

This is using falco 0.8.1 on Ubuntu Trusty LTS.

**Comment by JPLachance** [comment]
I have the same issue with my Kubernetes deployment. Here are some details:

Nodes:

- Kernel Version: 4.4.78-k8s
- OS Image: Debian GNU/Linux 8 (jessie)
- Operating System: linux
- Architecture: amd64
- Container Runtime Version: docker://1.12.6
- Kubelet Version: v1.8.2
- Kube-Proxy Version: v1.8.2

My Falco container:
- Image: sysdig/falco:0.8.1
- Last State: Terminated
- Reason: Error
- Exit Code: 139
- Started: Fri, 01 Dec 2017 10:16:38 -0500
- Finished: Fri, 01 Dec 2017 10:55:07 -0500  

And an exemple of the logs I get on the node:
```Dec  1 00:31:03 kernel: [14784.705391] falco[9184]: segfault at 7f44210f1d75 ip 00007f44225e7ff0 sp 00007ffd57725d48 error 4 in libanl-2.24.so[7f44225e7000+3000]```

I have let my Kubernetes DaemonSet running for 12h and one of the pods rebooted about 12 times. I have almost the default configuration (removed duplicated log outputs) and the default rule set.

I hope it helps!

**Comment by mstemm** [comment]
If you happen to narrow it down to a trace file that triggers the crash that would be very helpful in tracking down the problem. Otherwise can you capture a core dump/stack trace? You could try installing the dev build, which is compiled less aggressively.

**Comment by JPLachance** [comment]
I finally understood how to launch Falco in gdb in Kubernetes. Here is the resulting trace:
```bash
[...]
Building module:
cleaning build area...
make -j4 KERNELRELEASE=4.4.78-k8s -C /lib/modules/4.4.78-k8s/build M=/var/lib/dkms/falco/0.1.2338dev/build...........
cleaning build area...

DKMS: build completed.
[...]
DKMS: install completed.
* Trying to load a system falco-probe, if present
* Trying to load a dkms falco-probe, if present
falco-probe found and loaded in dkms
GNU gdb (Debian 7.12-6+b1) 7.12.0.20161007-git
[...]
Reading symbols from /usr/bin/falco...(no debugging symbols found)...done.
Starting program: /usr/bin/falco -K /var/run/secrets/kubernetes.io/serviceaccount/token -k https://kubernetes.default -pk
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
Mon Dec 11 21:47:49 2017: Falco initialized with configuration file /etc/falco/falco.yaml
Mon Dec 11 21:47:50 2017: Parsed rules from file /etc/falco/falco_rules.yaml
Mon Dec 11 21:47:50 2017: Parsed rules from file /etc/falco/falco_rules.local.yaml
[New Thread 0x7ffff7fe9780 (LWP 1152)]
[New Thread 0x7ffff1d05700 (LWP 1153)]
[Thread 0x7ffff7fe9780 (LWP 1152) exited]
[New Thread 0x7ffff1504700 (LWP 1154)]
[Thread 0x7ffff1d05700 (LWP 1153) exited]
[...]
[New Thread 0x7ffff7fe9780 (LWP 1168)]
[New Thread 0x7ffff1504700 (LWP 1169)]
[Thread 0x7ffff7fe9780 (LWP 1168) exited]
[Thread 0x7ffff1504700 (LWP 1169) exited]

Thread 1 "falco" received signal SIGSEGV, Segmentation fault.
0x00007ffff79cf050 in gai_error () from /lib/x86_64-linux-gnu/libanl.so.1
#0  0x00007ffff79cf050 in gai_error () from /lib/x86_64-linux-gnu/libanl.so.1
#1  0x000000000051ed4c in socket_data_handler<k8s_handler>::dns_cleanup(gaicb**) ()
#2  0x0000000000524897 in socket_data_handler<k8s_handler>::dns_cleanup() ()
#3  0x0000000000525fde in socket_data_handler<k8s_handler>::~socket_data_handler() ()
#4  0x0000000000515d9d in k8s_handler::check_collector_status() ()
#5  0x000000000051601b in k8s_handler::check_enabled() ()
#6  0x0000000000518310 in k8s_handler::collect_data() ()
#7  0x000000000052eab9 in k8s_net::watch() ()
#8  0x000000000058dd19 in sinsp::collect_k8s() ()
#9  0x00000000005907f8 in sinsp::update_k8s_state() ()
#10 0x0000000000591447 in sinsp::next(sinsp_evt**) ()
#11 0x00000000004acf8d in do_inspect(falco_engine*, falco_outputs*, sinsp*, unsigned long, std::string&) ()
#12 0x00000000004af7e4 in falco_init(int, char**) ()
#13 0x00007ffff6980561 in __libc_start_main ()
   from /lib/x86_64-linux-gnu/libc.so.6
#14 0x000000000049ed71 in _start ()
```

I removed the `-k https://kubernetes.default` argument and I did not get any SIGSEGV error since then.

**Comment by mstemm** [comment]
Thanks. We think this is related to fetching lower-level k8s information in sysdig's k8s module and there's a fix getting prepared now. Once it's been released, I'll release a patch of falco that includes the fix.

**Comment by mstemm** [comment]
This was a problem in the sysdig code, and was fixed in https://github.com/draios/sysdig/commit/3b558892c577782dc6539cd168556bb47a660f47#diff-ef0695fa60df23b4d886b6376cc935a9. The fix was included in falco 0.9.0. We didn't have an issue to track this within sysdig, although we did have an issue to track it internally for the sysdig monitor agent.

In any case, this crash should be fixed now. I'll close this issue, and if you see it again, please go ahead and reopen/file a new issue.
